### ISSUE TITLE
Duplication of packets seen in bridge-promiscuous mode when accessing a pod on the same node through service ip

### ISSUE THREAD

**Issue by rancohen1** [kubernetes/kubernetes#25793]
using 1.2.4

I created a pod with a small udp echo server running on port 5005 (mendhak/udp-listener) and exposed it as a udp service.

**This only happens when the client and server land on the same node!**

A simple client
```
nc -4u -w1 $SERVICE_IP 5005 <<< a
nc -4u -w1 $POD_IP 5005 <<< b
```
Meanwhile, on the server
```
Listening on UDP port 5005
a
a
b
```
tcpdump on the node
```
08:00:38.574388 IP 10.228.1.5.38130 > 10.228.1.3.5005: UDP, length 7
08:00:38.574447 IP 10.228.1.5.38130 > 10.228.1.3.5005: UDP, length 7
08:00:42.022864 IP 10.228.1.5.42386 > 10.228.1.3.5005: UDP, length 7
```
By the way, my real use case was using logstash with a udp input and my logs were duplicated sometimes (when the pod that produces logs landed on the same node as logstash).

another thought, maybe TCP packets get duplicated too but handled?

**Comment by lavalamp** [comment]
@thockin to either answer or give the "no support" boilerplate, I'm not sure if this is expected or not.

**Comment by thockin** [comment]
@girishkalele

**Comment by girishkalele** [comment]
@rancohen1 

Your observation about the intermittent duplication of packets only in the case of same node + service IP is the same as we observed recently. There is intermittent duplication of TCP packets too, mostly the smaller TCP ACKs.
The root cause is due to the bridge cbr0 being configured in promiscuous mode. This was introduced to workaround a kernel bug https://github.com/kubernetes/kubernetes/issues/20096

There is a knob HAIRPIN_MODE in the kube-up scripts to configure networking in bridge-promiscuous mode (which is the default today) vs the correct knob which would be to set hairpin-veth mode.

In hairpin-veth mode, you wouldn't see duplicate packets but you could see kernel hung issues.

Related PR
https://github.com/kubernetes/kubernetes/pull/21326

**Comment by girishkalele** [comment]
cc@bprashanth

**Comment by girishkalele** [comment]
The fix for this issue (changing default hairpin_mode to hairpin_veth) is gated by the fix for the kernel hosed issue; without the kernel fix, soak tests running hairpin_veth mode consistently are seeing hung nodes.

**Comment by girishkalele** [comment]
Closed via #28650
